src/software/jetson_nano/primitive_executor.cpp

```diff
@@ -1,54 +1,41 @@
-                                       const TbotsProto::Primitive& primitive)
+PrimitiveExecutor::PrimitiveExecutor(const double time_step,
+                                     const RobotConstants_t& robot_constants)
+    : current_primitive_(),
+      robot_constants_(robot_constants),
+      hrvo_simulator(static_cast<float>(time_step), robot_constants)
 {
-    robot_constants_   = robot_constants;
-    current_primitive_ = primitive;
 }
 
-Vector PrimitiveExecutor::getTargetLinearVelocity(
-    const TbotsProto::MovePrimitive& move_primitive, const RobotState& robot_state)
+void PrimitiveExecutor::updatePrimitiveSet(
+    const unsigned int robot_id, const TbotsProto::PrimitiveSet& primitive_set_msg)
 {
-    const float LOCAL_EPSILON = 1e-6f;  // Avoid dividing by zero
-
-    // Unpack current move primitive
-    auto destination              = move_primitive.path().point().at(0);
-    const float dest_linear_speed = move_primitive.final_speed_m_per_s();
-    const float max_speed_m_per_s = move_primitive.max_speed_m_per_s();
-    const Point final_position    = Point(destination.x_meters(), destination.y_meters());
-    std::clamp(max_speed_m_per_s, 0.0f, robot_constants_.robot_max_speed_m_per_s);
-
-    const float max_target_linear_speed = fmaxf(max_speed_m_per_s, dest_linear_speed);
-
-    // Compute distance to destination
-    const float norm_dist_delta =
-        static_cast<float>((robot_state.position() - final_position).length());
-
-    // Compute at what linear distance we should start decelerating
-    // d = (Vf^2 - Vi^2) / (2a + LOCAL_EPSILON)
-    const float start_linear_deceleration_distance =
-        (max_target_linear_speed * max_target_linear_speed -
-         dest_linear_speed * dest_linear_speed) /
-        (2 * robot_constants_.robot_max_acceleration_m_per_s_2 + LOCAL_EPSILON);
-
-    // When we are close enough to start decelerating, we reduce the max speed
-    // by 60%. Once we get closer than 0.6 meters, we start to linearly decrease
-    // speed proportional to the distance to the destination. 0.6 was determined
-    // experimentally.
-    float target_linear_speed = max_target_linear_speed;
-    if (norm_dist_delta < start_linear_deceleration_distance)
+    hrvo_simulator.updatePrimitiveSet(primitive_set_msg);
+    auto primitive_set_msg_iter = primitive_set_msg.robot_primitives().find(robot_id);
+    if (primitive_set_msg_iter != primitive_set_msg.robot_primitives().end())
     {
-        target_linear_speed = max_target_linear_speed * fminf(norm_dist_delta, 0.6f);
+        current_primitive_ = primitive_set_msg_iter->second;
     }
+}
+
+void PrimitiveExecutor::updateWorld(const TbotsProto::World& world_msg)
+{
+    hrvo_simulator.updateWorld(World(world_msg));
+}
 
-    Vector target_global_velocity = final_position - robot_state.position();
+Vector PrimitiveExecutor::getTargetLinearVelocity(const unsigned int robot_id,
+                                                  const RobotState& robot_state)
```

Comment:
seems a bit weird that hrvo doesn't know the orientation of the robots and it needs to be passed it, same with getTargetAngularVelocity

Reply:
HRVO does not care about orientation, but the orientation is required to calculate the local robot velocity. And `getTargetAngularVelocity` does not use HRVO and the implementation is the same as before. 
Could you elaborate on what you mean by it being weird?

Reply:
I think that clears things up for me that hrvo is a limited simulator. Perhaps we could pass in only the orientation into this function and not the rest of the robot state?

Reply:
Good point, updated to take in orientation in this and two other methods.